lib/spack/spack/directives.py

```diff
@@ -317,34 +317,95 @@ def remove_directives(arg):
+    hg: Optional[str] = None,
+    cvs: Optional[str] = None,
+    get_full_repo: Optional[bool] = None,
+    branch: Optional[str] = None,
+    submodules: Optional[bool] = None,
+    submodules_delete: Optional[bool] = None,
+    commit: Optional[str] = None,
+    tag: Optional[str] = None,
+    revision: Optional[str] = None,
+    date: Optional[str] = None,
+    # unsafe checksums
+    md5: Optional[str] = None,
+    sha1: Optional[str] = None,
+    sha224: Optional[str] = None,
+):
     """Adds a version and, if appropriate, metadata for fetching its code.
 
     The ``version`` directives are aggregated into a ``versions`` dictionary
     attribute with ``Version`` keys and metadata values, where the metadata
     is stored as a dictionary of ``kwargs``.
 
-    The ``dict`` of arguments is turned into a valid fetch strategy for
+    The (keyword) arguments are turned into a valid fetch strategy for
     code packages later. See ``spack.fetch_strategy.for_package_version()``.
-
-    Keyword Arguments:
-        deprecated (bool): whether or not this version is deprecated
     """
 
     def _execute_version(pkg):
-        if checksum is not None:
-            if hasattr(pkg, "has_code") and not pkg.has_code:
-                raise VersionChecksumError(
-                    "{0}: Checksums not allowed in no-code packages"
-                    "(see '{1}' version).".format(pkg.name, ver)
-                )
+        if (
+            any((sha256, sha384, sha512, md5, sha1, sha224, checksum))
+            and hasattr(pkg, "has_code")
+            and not pkg.has_code
+        ):
+            raise VersionChecksumError(
+                "{0}: Checksums not allowed in no-code packages "
+                "(see '{1}' version).".format(pkg.name, ver)
+            )
 
-            kwargs["checksum"] = checksum
+        kwargs = {
+            key: value
+            for key, value in (
+                ("sha256", sha256),
+                ("sha384", sha384),
+                ("sha512", sha512),
+                ("preferred", preferred),
+                ("deprecated", deprecated),
+                ("expand", expand),
+                ("url", url),
+                ("extension", extension),
+                ("no_cache", no_cache),
+                ("fetch_options", fetch_options),
+                ("git", git),
+                ("svn", svn),
+                ("hg", hg),
+                ("cvs", cvs),
+                ("get_full_repo", get_full_repo),
+                ("branch", branch),
+                ("submodules", submodules),
+                ("submodules_delete", submodules_delete),
+                ("commit", commit),
+                ("tag", tag),
+                ("revision", revision),
+                ("date", date),
+                ("md5", md5),
+                ("sha1", sha1),
+                ("sha224", sha224),
+                ("checksum", checksum),
+            )
+            if value is not None
+        }
```

Comment:
Ideally we wouldn't maintain this twice:

```suggestion
        code = version.__code__
        start = code.co_argcount
        names = code.co_varnames[start:start + code.co_kwonlyargcount]
        
        values = locals()
        kwargs = {name: values[name] for name in names if values[name] is not None}
```

Only worry with this is that it might make `version()` slow -- can you see if it does?  Timing this should give you an idea:

```console
time spack python -c 'import spack.repo; classes = list(spack.repo.path.all_package_classes())'
```

Reply:
I thought about it, but I'm not a big fan of introspection and I don't think there's going to be a whole lot of maintenance.

Reply:
I think you should try it

Reply:
I get the point, but I don't agree it's an improvement to maintainability. For example `version.__code__` refers to the decorator, not the decorated function, and the decorator has no key-word only arguments. Also `checksum` is a positional or keyword argument, not keyword only, and still needs to be added to the dict, but the other positional argument shouldn't. On top of that static analysis doesn't capture this and worst case it or your editor will complain about unused keyword arguments.

Regarding performance it's +5% min time of 10 runs, without fixing the bug ^, so it's a lower bound given that `names` was empty.

Reply:
Ok if it adds even a bit of time let’s get rid of it. 5% is too much. I *wish* we did not have to write this twice but I don’t see a great way around it, other than using kwargs.

It’s kind of too bad there isn’t a good way to still refer to the dict for delegation, but oh well.